pandas/tests/io/excel/test_writers.py

```diff
@@ -1166,7 +1173,9 @@ def test_bytes_io(self, engine):
         writer.save()
 
         bio.seek(0)
-        reread_df = pd.read_excel(bio, index_col=0)
+        if engine != "odf":
+            engine = None
```

Comment:
Are you setting to None because this test doesn't work for odf? If so you should just `pytest.xfail()` if not implemented or `pytest.skip()` if not applicable

Reply:
No, when you set engine to None, it uses the default engine (xlrd IIRC)

Reply:
Yea so my point is we shouldn't be changing the fixture like this. Should either be an xfail or skip depending on comment above

Reply:
No tests are skipped, only the right engine is chosen. When having a byte iostream you have to explicitly pass the engine, or if you pass `None` it uses the default engine (which is `xlrd` IIRC).

Reply:
Can you take a look at test_readers to see how we handle this? Can ideally ensure we parametrize with the appropriate writer / extension combinations up front; see `engine_and_read_ext` in test_readers

Reply:
I have now implemented an automatic file type recognizer for byte streams. Currently the only choice to be made is between 'xlrd' and 'odf', so it is quite easy.
For the other test I have improved the extension recognition to more types of `path`.
Plus I have fixed the parameterization of the tests that depend on the various external Excel read and write packages.